cmakebuild/docs/man/man3/m_verify.3
.TH "m_verify" 3 "Tue Feb 20 2018" "Mstdlib-1.0.0" \" -*- nroff -*-
.ad l
.nh
.SH NAME
m_verify \- Validation of strings against various standards (email address, domain name, etc\&.)\&.  

.SH SYNOPSIS
.br
.PP
.SS "Enumerations"

.in +1c
.ti -1c
.RI "enum \fBM_verify_email_listdelim_t\fP { \fBM_VERIFY_EMAIL_LISTDELIM_AUTO\fP = 0, \fBM_VERIFY_EMAIL_LISTDELIM_SEMICOLON\fP, \fBM_VERIFY_EMAIL_LISTDELIM_COMMA\fP }"
.br
.in -1c
.SS "Functions"

.in +1c
.ti -1c
.RI "M_bool \fBM_verify_email_address\fP (const char *addr)"
.br
.ti -1c
.RI "M_bool \fBM_verify_domain\fP (const char *dname)"
.br
.ti -1c
.RI "M_bool \fBM_verify_email_address_list\fP (const char *addr_list, \fBM_verify_email_listdelim_t\fP delim_type)"
.br
.in -1c
.SH "Detailed Description"
.PP 
Validation of strings against various standards (email address, domain name, etc\&.)\&. 


.SH "Enumeration Type Documentation"
.PP 
.SS "enum \fBM_verify_email_listdelim_t\fP"

.PP
\fBEnumerator\fP
.in +1c
.TP
\fB\fIM_VERIFY_EMAIL_LISTDELIM_AUTO \fP\fP
.TP
\fB\fIM_VERIFY_EMAIL_LISTDELIM_SEMICOLON \fP\fP
.TP
\fB\fIM_VERIFY_EMAIL_LISTDELIM_COMMA \fP\fP
.SH "Function Documentation"
.PP 
.SS "M_bool M_verify_email_address (const char * addr)"
Verify that the given single email address is valid\&.
.PP
An email address is composed of the following parts: 
.PP
.nf
display_name<local_part@domain>
.fi
.PP
.PP
The display name may contain double-quotes\&. Within double-quotes, the full range of symbols may be used (dot, comma, semicolon, at, escaped double-quote, etc\&.)\&. The space and tab characters are allowed in the display name, even outside of quotes\&.
.PP
If the display name is not provided, the angle brackets around the rest of the email address are optional\&.
.PP
This email address validator differs from the RFC standards for email in the following ways:
.PP
.PD 0
.IP "\(bu" 2
Double quotes are not allowed, except in the display name (a\&.k\&.a\&. friendly name)\&. 
.IP "\(bu" 2
Comments (chunks of text enclosed by parentheses) are not allowed, though parenthesis are allowed in the display name if they're surrounded by double-quotes\&. 
.IP "\(bu" 2
Consecutive dots are allowed in the local part of the address, since they are accepted by gmail\&. 
.IP "\(bu" 2
Only the following non-alphanumeric characters are allowed in the local part:
.PP
.nf
'*+-=^_{}.~ 
.fi
.PP
 
.IP "\(bu" 2
IP address domains are not allowed (e\&.g\&., joe\&.smith@[192\&.168\&.2\&.1] is invalid)\&. 
.IP "\(bu" 2
UTF-8 is not allowed in any part of the address (display name, local part, or domain name)\&.
.PP
Example of valid address: 
.PP
.nf
"Public, John Q. (Support)" <jq..public+support@generic-server.com> 
.fi
.PP
.PP
\fBSee also:\fP
.RS 4
\fBM_verify_email_address_list\fP
.RE
.PP
\fBParameters:\fP
.RS 4
\fIaddr\fP email address to verify 
.RE
.PP
\fBReturns:\fP
.RS 4
TRUE if the address is valid 
.RE
.PP

.SS "M_bool M_verify_domain (const char * dname)"
Verify that the given domain name is valid\&.
.PP
Valid domain names are composed of letters, digits, and hyphens ('-'), separated into sections by dots ('\&.')\&. Each section is limited to 63 chars, and may not begin or end with a hyphen\&. The entire domain name is limited to 253 characters - note that a 253 character domain name takes 255 bytes to store with the length-prefixed encoding used by DNS\&.
.PP
\fBParameters:\fP
.RS 4
\fIdname\fP domain name to verify 
.RE
.PP
\fBReturns:\fP
.RS 4
TRUE if the domain name is valid 
.RE
.PP

.SS "M_bool M_verify_email_address_list (const char * addr_list, \fBM_verify_email_listdelim_t\fP delim_type)"
Verify that the given list of email addresses is valid\&.
.PP
The list may be delimited by either commas (',') or semicolons (';'), but not both\&. This function limits the number of addresses in the list to 100 - this isn't a true standard, but it's the limit used virutally everywhere in practice\&.
.PP
The list parser is tolerant of whitespace - individual addresses are trimmed before they're passed to the email address validator\&.
.PP
The individual addresses in the list may include display names (a\&.k\&.a friendly names)\&.
.PP
\fBSee also:\fP
.RS 4
\fBM_verify_email_address\fP
.RE
.PP
\fBParameters:\fP
.RS 4
\fIaddr_list\fP list of email addresses to verify 
.br
\fIdelim_type\fP Type of delimiter to expect\&. Can be auto detected\&. 
.RE
.PP
\fBReturns:\fP
.RS 4
TRUE if the list and all emails in it are valid 
.RE
.PP

.SH "Author"
.PP 
Generated automatically by Doxygen for Mstdlib-1\&.0\&.0 from the source code\&.


formats/verify/m_verify.c
/* The MIT License (MIT)
 * 
 * Copyright (c) 2015 Main Street Softworks, Inc.
 * 
 * Permission is hereby granted, free of charge, to any person obtaining a copy
 * of this software and associated documentation files (the "Software"), to deal
 * in the Software without restriction, including without limitation the rights
 * to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
 * copies of the Software, and to permit persons to whom the Software is
 * furnished to do so, subject to the following conditions:
 * 
 * The above copyright notice and this permission notice shall be included in
 * all copies or substantial portions of the Software.
 * 
 * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
 * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
 * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
 * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
 * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
 * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
 * THE SOFTWARE.
 */

#include "m_config.h"

#include <mstdlib/mstdlib.h>
#include <mstdlib/mstdlib_formats.h>


#define MAX_LEN_EMAIL        254 /* Max. number of chars in an entire email address */

#define MAX_LEN_EMAIL_LOCAL  64  /* Max. number of chars in local part of email (the part before the '@'). */

#define MAX_LEN_DOMAIN       253 /* Max. number of chars in a domain name (255 bytes, but 253 chars due to encoding)*/

#define MAX_LEN_DNS_LABEL    63  /* Max. number of chars between dots in a domain name. */

#define MAX_EMAIL_RECIPIENTS 100 /* Max. number of recipients for a single email (not standard, but pretty universal) */


/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

static const char *allowed_in_local =
	"0123456789"                 /*[0-9]*/
	"abcdefghijklmnopqrstuvwxyz" /*[a-z]*/
	"ABCDEFGHIJKLMNOPQRSTUVWXYZ" /*[A-Z]*/
	".'*+-=^_{}~";               /* Symbols list more restrictive than standard, to avoid common conflicts
								  * with symbols used in shell scripts andweird protocols that have been
								  * grafted on top of email over the years.
								  */

static const char *allowed_in_dns_label =
	"0123456789"
	"abcdefghijklmnopqrstuvwxyz"
	"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	"-";

static const char *allowed_in_display_name =
	"0123456789"
	"abcdefghijklmnopqrstuvwxyz"
	"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	"!#$%&'*+-/=?^_`{|}~ \t";

static const char *allowed_in_quoted_display_name =
	"0123456789"
	"abcdefghijklmnopqrstuvwxyz"
	"ABCDEFGHIJKLMNOPQRSTUVWXYZ"
	"!#$%&'*+-/=?^_`{|}~ \t"
	"()<>[]:;@,.";               /* these are only allowed if inside quotes */


static M_bool is_escaped(const char *str, const char *str_pos, char escape)
{
	size_t escape_count;

	/* NOTE: an escape character may escape itself. So, we have to count the number of escape characters
	 *       before the current character, in order to determine if this character is escaped.
	 * 
	 * Ex: \\\" --> " is escaped by \
	 *     \\"  --> " is not escaped
	 *     \"   --> " is escaped by \
	 */

	/* Count the number of escapes before the current character. */
	escape_count = 0;
	while (str_pos > str && *(str_pos - 1) == escape) {
		escape_count++;
		str_pos--;
	}

	/* If the current char is preceeded by an odd number of escapes, this character is escaped. */
	return (escape_count % 2 == 1)? M_TRUE : M_FALSE;
}


/* Count the number of times 'ch' appears in 'str', not including instances
 * where 'ch' is inside quotes.
 */
static size_t str_count_chars_quoted(const char *str, char ch, char quote, char escape)
{
	const char *str_pos   = str;
	size_t      count     = 0;
	M_bool      in_quotes = M_FALSE;

	if (str == NULL) {
		return 0;
	}

	while (*str_pos != '\0') {
		if (*str_pos == quote && !is_escaped(str, str_pos, escape)) {
			in_quotes = !in_quotes; /* toggle */
		} else if (*str_pos == ch && !in_quotes) {
			count++;
		}
		str_pos++;
	}

	return count;
}


/* Possible formats (all valid):
 * (1) user@example.com
 * (2) <user@example.com>
 * (3) John User <user@example.com>
 * (4) "John Q. User" <user@example.com>
 * (5) "<noise> (junk) []:;@..." <user@example.com>
 */
static M_bool verify_display_name(const char *disp)
{
	M_bool  ret            = M_TRUE;
	char   *quoted_disp;
	char   *unquoted_disp;


	/* Divide content between quoted and unquoted parts. */
	quoted_disp   = M_str_keep_quoted(disp, '"', '\\');
	unquoted_disp = M_str_remove_quoted(disp, '"', '\\');

	if (!M_str_isempty(quoted_disp) && !M_str_ischarset(quoted_disp, allowed_in_quoted_display_name)) {
		ret = M_FALSE;
	}

	if (!M_str_isempty(unquoted_disp) && !M_str_ischarset(unquoted_disp, allowed_in_display_name)) {
		ret = M_FALSE;
	}

	M_free(quoted_disp);
	M_free(unquoted_disp);

	return ret;
}


static M_bool verify_local_part(const char *local_part)
{
	size_t len = M_str_len(local_part);

	if (len == 0 || len > MAX_LEN_EMAIL_LOCAL) {
		return M_FALSE;
	}

	/* Can't start or end with a dot. */
	if (local_part[0] == '.' || local_part[len - 1] == '.') {
		return M_FALSE;
	}

	/* Can't start with a hyphen. */
	if (local_part[0] == '-') {
		return M_FALSE;
	}

	/* Only allowed to contain characters from "allowed_in_local". */
	return M_str_ischarset(local_part, allowed_in_local);
}


static M_bool verify_dns_label(const char *dns_label)
{
	size_t len = M_str_len(dns_label);

	if (len == 0 || len > MAX_LEN_DNS_LABEL) {
		return M_FALSE;
	}

	/* Can't start or end with a hyphen. */
	if (dns_label[0] == '-' || dns_label[len - 1] == '-') {
		return M_FALSE;
	}

	/* Only allowed to contain characters from "allowed_in_dns_label". */
	return M_str_ischarset(dns_label, allowed_in_dns_label);
}


/* - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - */

M_bool M_verify_domain(const char *dname)
{
	size_t   num_dns_labels = 0;
	char   **dns_labels;
	size_t   i;

	if (M_str_isempty(dname) || M_str_len(dname) > MAX_LEN_DOMAIN) {
		return M_FALSE;
	}

	/* Don't allow comments (in parenthesis) or IP addresses (in brackets). */

	/* Split by dot, keep empty list elements. */
	dns_labels = M_str_explode_str('.', dname, &num_dns_labels);

	for (i=0; i<num_dns_labels; i++) {
		if (!verify_dns_label(dns_labels[i])) {
			M_str_explode_free(dns_labels, num_dns_labels);
			return M_FALSE;
		}
	}

	M_str_explode_free(dns_labels, num_dns_labels);
	return M_TRUE;
}


M_bool M_verify_email_address(const char *addr)
{
	char    *email            = NULL;
	char    *display_name     = NULL;

	size_t   num_email_parts  = 0;
	char   **email_parts      = NULL;

	size_t   len;
	size_t   num_open_angle;
	size_t   num_close_angle;

	M_bool   ret              = M_FALSE;


	len = M_str_len(addr);
	if (len < 3 || len > MAX_LEN_EMAIL) {
		goto done;
	}

	/* Split string between the email address itself, and the display name (if any).
	 *
	 * Possible formats (all valid):
	 * (1) user@example.com
	 * (2) <user@example.com>
	 * (3) John User <user@example.com>
	 * (4) "John Q. User" <user@example.com>
	 * (5) "<noise> (junk) []:;@..." <user@example.com>
	 */
	num_open_angle  = str_count_chars_quoted(addr, '<', '"', '\\');
	num_close_angle = str_count_chars_quoted(addr, '>', '"', '\\');
	if (num_open_angle != num_close_angle || num_open_angle > 1) {
		/* Make sure there's one pair of unquoted angle brackets, at the most. */
		goto done;
	}
	if (num_open_angle > 0) {
		if(addr[len - 1] != '>') {
			/* If the email address is surrounded by brackets, make sure it's the last thing in the string. */
			goto done;
		}
		email        = M_str_keep_bracketed_quoted(addr, '<', '>', '"', '\\');
		display_name = M_str_remove_bracketed_quoted(addr, '<', '>', '"', '\\');
	} else {
		email        = M_strdup(addr);
	}


	/* Validate display name. */
	if (!M_str_isempty(display_name) && !verify_display_name(display_name)) {
		goto done;
	}


	/* Validate email. */
	/* -- Split by @ sign. Valid addresses must have one, and exactly one, @ sign, since we don't allow quoting. */
	email_parts = M_str_explode_str('@', email, &num_email_parts);
	if (num_email_parts != 2) {
		goto done;
	}
	if (!verify_local_part(email_parts[0])) {
		goto done;
	}
	if (!M_verify_domain(email_parts[1])) {
		goto done;
	}


	/* If we reach here, the address was valid. */
	ret = M_TRUE;

	done:
	M_str_explode_free(email_parts, num_email_parts);
	M_free(display_name);
	M_free(email);
	return ret;
}


M_bool M_verify_email_address_list(const char *addr_list, M_verify_email_listdelim_t delim_type)
{
	size_t          num_addresses  = 0;
	char          **addresses      = NULL;
	size_t          num_commas;
	size_t          num_semicolons;
	unsigned char   delimiter      = ',';
	size_t          i;

	if (M_str_isempty(addr_list)) {
		return M_FALSE;
	}

	/* We do a trim later on that eats all six whitespace chars, but we only want to
	 * allow space and tab. So, make sure none of the other four whitespace chars
	 * are present in the address list ahead of time.
	 */
	if (!M_str_isnotcharset(addr_list, "\r\n\f\v")) {
		return M_FALSE;
	}

	/* List may be delimited either by commas or semicolons (but not both). */
	num_commas     = str_count_chars_quoted(addr_list, ',', '"', '\\');
	num_semicolons = str_count_chars_quoted(addr_list, ';', '"', '\\');
	switch (delim_type) {
		case M_VERIFY_EMAIL_LISTDELIM_AUTO:
			if (num_commas > 0 && num_semicolons > 0) {
				return M_FALSE;
			} else if (num_commas > 0) {
				delimiter = ',';
			} else if (num_semicolons > 0) {
				delimiter = ';';
			}
			break;
		case M_VERIFY_EMAIL_LISTDELIM_COMMA:
			delimiter = ',';
			break;
		case M_VERIFY_EMAIL_LISTDELIM_SEMICOLON:
			delimiter = ';';
			break;
	}

	/* If there are no delimiters, evalute the string as a single address. */
	if ((delimiter == ',' && num_commas == 0) ||
		(delimiter == ';' && num_semicolons == 0))
	{
		return M_verify_email_address(addr_list);
	}

	/* Split by the detected delimiter, the check each individual address (after whitespace trimming).
	 * Don't allow empty list entries, fail if we detect one (including entries that are all whitespace).
	 */
	addresses = M_str_explode_str_quoted(delimiter, addr_list, '"', '\\', 0, &num_addresses);
	if (num_addresses > MAX_EMAIL_RECIPIENTS) {
		M_str_explode_free(addresses, num_addresses);
		return M_FALSE;
	}
	for (i=0; i<num_addresses; i++) {
		char *address = M_str_trim(addresses[i]);
		if (M_str_isempty(address) || !M_verify_email_address(address)) {
			M_str_explode_free(addresses, num_addresses);
			return M_FALSE;
		}
	}

	M_str_explode_free(addresses, num_addresses);
	return M_TRUE;
}
